enhancements/network/ovs-hardware-offload.md

```diff
@@ -0,0 +1,242 @@
+enables offloading of OVS datapath rules to a compatible network card using
+linux Traffic Control (TC) tool `tc`. It relies on SR-IOV technology to control
+the SR-IOV Virtual Functions (VF) via VF representor port that is plugged
+into OpenvSwitch bridge. It improves the utilization of available bandwidth
+on underlying NICs by moving CPU intensive network packet processing from
+OVS software to hardware.
+
+## Motivation
+
+Enable high-performance datapath on primary pod interface with production-ready
+OVN-Kubernetes.
+
+### Goals
+
+Enable OpenFlow hardware offload for OVN datapath
+
+### Non-Goals
+
+Replace OpenFlow software based solution
+
+## Proposal
+
+OVS Hardware Offload accelerates the OVS software flows with a SR-IOV NIC
+that works in `switchdev` mode. It requires configurations in both OpenvSwitch
+daemon and SR-IOV hardware NIC.
+
+- OpenvSwitch daemon is managed by Cluster Network Operator (CNO) and runs in
+`ovs-node` DaemonSet. A boolean configurable field will be added in CNO
+`Network` Custom Resource Definition `spec.defaultNetwork.ovnKubernetesConfig`
+to enable or disable OVS Hardware Offload feature for OpenvSwitch daemon.
+
+- SR-IOV hardware NIC is managed by SR-IOV Network Operator and will be
+provisioned with a new device type `switchdev`. The initial supported NIC
+model will be Mellanox MT27800 Family [ConnectX-5] 25GbE dual-port SFP28.
+
+Besides CNO and SR-IOV Operator changes, there are several other dependencies.
+
+- OVS Hardware Offload will only be supported on bare metal deployment.
+
+- Multus supports overwriting cluster network definition by specifying a
+customized default net-attach-def in pod annotation. Customized net-attach-def
+will still use `ovn-k8s-cni-overlay` CNI plugin, but with an special annotation
+that requests a SR-IOV device.
+
+- RHCOS image shall be based on RHEL-8.3 for fully support of OVS Hardware
+Offload with connection track feature.
+
+- OpenvSwitch package version 2.13.
+
+- Mellanox SR-IOV VF-LAG provides hardware link-aggregation (LAG) for hardware
+offloaded ports. It allows offloading of OVS rules to a linux `bond` network
+device that combines two idential and hardware offload capable Physical
+Functions (PFs). In order to use SR-IOV VF-LAG, a bond interface needs to be
+created and used as geneve endpoint of OVS bridge. With the SR-IOV VF LAG
+capability, the NIC PFs can receive the rules that the OVS tries to offload to
+the linux bond net device and offload them to the hardware e-switch.
+
+#### RHCOS
+
+- RHEL 8.3 be used as RHCOS image base for fully support of OVS Hardware Offload
+with connection track feature.
+
+#### OpenvSwitch
+
+- OpenvSwitch package [version 2.13](https://github.com/openshift/ovn-kubernetes/pull/122) is currently used in OpenShift.
+
+#### OVN-Kubernetes
+
+- OVS Hardware Offload is a [feature available](https://github.com/ovn-org/ovn-kubernetes/blob/master/docs/ovs_offload.md) in OVN-Kubernetes upstream.
```

Comment:
Thanks for sharing the `gateway` discussion link!
What I'd like to propose is to use `local-gateway` mode that OCP is currently using.
I did some test with current ocp 4.5 version on baremetal, I could see that OVS tries to offload the flows to hardware NIC, but failed due to `Operation not supported`.  Current investigation shows that it might be due to conntrack feature is not available in underlying kernel version which is RHEL-8.1, but I cannot exclude atm the possibility that it's due to ovs hw offload doesn't work with `local-gateway` mode.

I did have some discussion with OpenStack and RHEL folks on whether ovs hw offload works in `local-gateway` or `shared-gateway` modes and the feedback so far are consistent and postive. @krsacme could help confirm this.

The network topologies discussed were (bothall works with ovs hw offload):
1) ovs hw offload on primary PF directly (local gateway)
![OVS HW OFFLOAD  -- Primary NIC PF](https://user-images.githubusercontent.com/6268701/80444941-9a23af00-8945-11ea-93dc-a3ba34bfa24f.jpg)

2) ovs hw offload on bond (shared-gateway)
![OVS-HW-OFFLOAD -- Bond](https://user-images.githubusercontent.com/6268701/80444975-b45d8d00-8945-11ea-9f8d-407a03373f38.jpg)

3) ovs hw offload on vlan on bond (local gateway)
![OVS HW OFFLOAD -- Primary NIC bond + vlan](https://user-images.githubusercontent.com/6268701/80445015-d0612e80-8945-11ea-94d4-c94b4c628998.png)

Reply:
So, I don't have nice color diagrams, but the issue is that: in shared gateway mode, the entire path from the pod's `eth0` to the external network can be offloaded into the NIC. In local gateway mode, that's not true; traffic going from a pod to the external network has to pass through the node's root network namespace and have iptables rules applied to it, etc, and that part can't be offloaded to the NIC. So you can't just have the pod connected to a VF that's connected directly to the external network.

Reply:
I see, that's a solid point.  I didn't realize the iptable rules applied to traffic going out to external network. Looks like `shared gateway` mode is what's required for offloading entire path from ovs to hw.
I'll update the doc to add the `shared gateway` mode as dependency and check if there is solution that dealing with the iptable path offloading.
btw, do you see any issue for internal traffic using ovs hw offoad? can we still make the statement that ovs hw offload works for pod-to-pod, service-to-service, pod-to-service traffic in current `local gateway` mode?

Reply:
The primary advantage of using offload is to get the flows to hardware so that the packet processing has higher throughput. Traffic directly to the VF of VM once the flow is offloaded (after first packet miss). I can confirm that in OpenStack we use the `shared gateway` (understood the terms with above explanation). There is no intermediate, so the complete packet processing can be offloaded.

Reply:
@zshi-redhat if it's possible to set things up so that pod-to-pod traffic gets fully-offloaded, but pod-to-external traffic is either only-partially-offloaded, or not offloaded at all, then that should work.

However, as discussed in https://github.com/ovn-org/ovn-kubernetes/issues/1193, we'd kind of like to move OCP to using shared-gateway mode anyway, eventually.

Reply:
@danwinship when hw-offload is enabled in ovs, I don't see there is a way to control what flows can be offloaded or not. If a pod may initiate external traffic, then user can choose to create pod with regular veth device instead of using SR-IOV device.
btw, for pod-to-external traffic, my understanding is that the not offloaded part includes iptable chains in nat table such as `OVN-KUBE-SNAT-MGMTPORT`, `OVN-KUBE-NODEPORT` in the host namespace, am I correct?

I agree that ovs hw offload shall be used with shared-gateway mode in longer term.

Reply:
ocp has moved to use shared-gateway mode